# Project 1: SAT & ACT Analysis

The first markdown cell in a notebook is a great place to provide an overview of your entire project. You will likely want to at least state your

## Problem Statement

Define the problem you are finding answers for from the data given.

## Executive Summary

If you want to, it's great to use relative links to direct your audience to various sections of a notebook. **HERE'S A DEMONSTRATION WITH THE CURRENT SECTION HEADERS**:

### Contents:
- [2017 Data Import & Cleaning](#Data-Import-and-Cleaning)
- [2018 Data Import and Cleaning](#2018-Data-Import-and-Cleaning)
- [Exploratory Data Analysis](#Exploratory-Data-Analysis)
- [Data Visualization](#Visualize-the-data)
- [Descriptive and Inferential Statistics](#Descriptive-and-Inferential-Statistics)
- [Outside Research](#Outside-Research)
- [Conclusions and Recommendations](#Conclusions-and-Recommendations)

**If you combine your problem statement, executive summary, data dictionary, and conclusions/recommendations, you have an amazing README.md file that quickly aligns your audience to the contents of your project.** Don't forget to cite your data sources!

*All libraries used should be added here*

In [1]:
#Imports:

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

sns.set_style('whitegrid')

## 2017 Data Import and Cleaning

#### 1. Read In SAT & ACT  Data

Read in the `sat_2017.csv` and `act_2017.csv` files and assign them to appropriately named pandas dataframes.

In [2]:
# Reading the CSV files using relative filepaths

sat17 = pd.read_csv("../data/sat_2017.csv")
act17 = pd.read_csv("../data/act_2017.csv")

#### 2. Display Data

Print the first 10 rows of each dataframe to your jupyter notebook

In [3]:
# Printing a few lines of each dataframe to ensure that it was read properly.

sat17.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            5%                                 593   
1                Alaska           38%                                 547   
2               Arizona           30%                                 563   
3              Arkansas            3%                                 614   
4            California           53%                                 531   
5              Colorado           11%                                 606   
6           Connecticut          100%                                 530   
7              Delaware          100%                                 503   
8  District of Columbia          100%                                 482   
9               Florida           83%                                 520   

   Math  Total  
0   572   1165  
1   533   1080  
2   553   1116  
3   594   1208  
4   524   1055  
5   595   1201  
6   512   1041  
7   492    996  
8   468    950  
9   497   1017

In [4]:
sat17.tail()

State Participation  Evidence-Based Reading and Writing  Math  \
46       Virginia           65%                                 561   541   
47     Washington           64%                                 541   534   
48  West Virginia           14%                                 558   528   
49      Wisconsin            3%                                 642   649   
50        Wyoming            3%                                 626   604   

    Total  
46   1102  
47   1075  
48   1086  
49   1291  
50   1230

In [5]:
# Checking the number of rows in the dataframe.

len(sat17)

51

In [6]:
# Checking the number of values in the 'State' column.

len(sat17['State'])

51

In [7]:
# Checking the unique values in the 'State' column.

sat17['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'District of Columbia',
       'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

In [8]:
act17.head(10)

State Participation  English  Math  Reading  Science  \
0              National           60%     20.3  20.7     21.4     21.0   
1               Alabama          100%     18.9  18.4     19.7     19.4   
2                Alaska           65%     18.7  19.8     20.4     19.9   
3               Arizona           62%     18.6  19.8     20.1     19.8   
4              Arkansas          100%     18.9  19.0     19.7     19.5   
5            California           31%     22.5  22.7     23.1     22.2   
6              Colorado          100%     20.1  20.3     21.2     20.9   
7           Connecticut           31%     25.5  24.6     25.6     24.6   
8              Delaware           18%     24.1  23.4     24.8     23.6   
9  District of Columbia           32%     24.4  23.5     24.9     23.5   

  Composite  
0      21.0  
1      19.2  
2      19.8  
3      19.7  
4      19.4  
5      22.8  
6      20.8  
7      25.2  
8      24.1  
9      24.2

In [9]:
act17.tail()

State Participation  English  Math  Reading  Science Composite
47       Virginia           29%     23.5  23.3     24.6     23.5      23.8
48     Washington           29%     20.9  21.9     22.1     22.0      21.9
49  West Virginia           69%     20.0  19.4     21.2     20.5      20.4
50      Wisconsin          100%     19.7  20.4     20.6     20.9      20.5
51        Wyoming          100%     19.4  19.8     20.8     20.6     20.2x

In [10]:
# Checking the number of rows in the dataframe.

len(act17)

52

In [11]:
# Checking the number of values in the 'State' column.

len(act17['State'])

52

In [12]:
# Checking the unique values in the 'State' column.

act17['State'].unique()

array(['National', 'Alabama', 'Alaska', 'Arizona', 'Arkansas',
       'California', 'Colorado', 'Connecticut', 'Delaware',
       'District of Columbia', 'Florida', 'Georgia', 'Hawaii', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Kansas', 'Kentucky', 'Louisiana',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Mississippi', 'Missouri', 'Montana', 'Nebraska', 'Nevada',
       'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Carolina', 'South Dakota',
       'Tennessee', 'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

#### 3. Verbally Describe Data

Take your time looking through the data and thoroughly describe the data in the markdown cell below. 

**SAT 2017:** According to the [source](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/), this dataset *(imported here as 'sat17')* provides a breakdown of the 2017 SAT participation rates *(column 'Participation')*, average scores for EBRW and Math sections *(columns 'Evidence-Based Reading and Writing' and 'Math', respectively)* and the average aggregate scores *(column 'Total')* for all 50 states and 1 federal district (District of Columbia) of the USA *(column 'State', arranged alphabetically)*.

The values in the 'State' column are *strings*. The numerical values of the scores in the dataset do not carry any decimal places, and so should be expected to be *integers*. The values in the 'Participation' column carry a numerical value followed by a % symbol. So, these should be expected to be *strings* which would need to be converted to *floats*.

**ACT 2017:** Similar to the above, according to the [source](https://www.act.org/content/dam/act/unsecured/documents/cccr2017/ACT_2017-Average_Scores_by_State.pdf), this dataset *(imported here as 'act17')* provides a breakdown of the 2017 ACT participation rates *(column 'Participation')*, average scores for English, Math, Reading and Science sections *(columns 'English', 'Math', 'Reading' and 'Science', respectively)* and the average composite scores *(column 'Composite')* for all 50 states and 1 federal district (District of Columbia) of the USA *(column 'State', arranged alphabetically)*. This dataset also contains one additional row of data for the National participation rate and average section & composite scores.

The values in the 'State' column are *strings*. The numerical values of the scores in the dataset carry decimal places, and so should be expected to be *floats*. The values in the 'Participation' column carry a numerical value followed by a % symbol. So, these should be expected to be *strings* which would need to be converted to *floats*.

#### 4a. Does the data look complete? 

In [13]:
# Checking for any empty values in SAT 2017 dataset.

sat17.isna().sum()

State                                 0
Participation                         0
Evidence-Based Reading and Writing    0
Math                                  0
Total                                 0
dtype: int64

In [14]:
# Checking for any empty values in ACT 2017 dataset.

act17.isna().sum()

State            0
Participation    0
English          0
Math             0
Reading          0
Science          0
Composite        0
dtype: int64

**Answer:** Yes, the data is complete as there are no empty values in the two datasets.

#### 4b. Are there any obvious issues with the observations?

**What is the minimum *possible* value for each test/subtest? What is the maximum *possible* value?**

Consider comparing any questionable values to the sources of your data:
- [SAT](https://blog.collegevine.com/here-are-the-average-sat-scores-by-state/)
- [ACT](https://blog.prepscholar.com/act-scores-by-state-averages-highs-and-lows)

**Answer:** *The answers in this section are ONLY based on a quick visual inspection of the CSV files. This method may be usable here because the datasets are relatively small in size (only 5-7 columns, ~51 rows).*

For the **SAT**, according to [this page](https://blog.prepscholar.com/how-is-the-sat-scored-scoring-charts), each of the **2 sections** have a 'scaled score' **range of 200 - 800 (inclusive)**. The total score is calculated by adding the 2 sections' scores, thus giving a possible **total score range of 400 - 1600**.

A quick visual inspection of the sat_2017 CSV file indicates that one of the values in the dataset is outside the above given ranges. This will have to be corrected, and all other values will also have to be checked using code, to ensure that they are within the correct ranges (which may have been missed during the visual inspection).

For the **ACT**, according to [this page](https://blog.prepscholar.com/how-is-the-act-scored), each of the **4 sections** have a 'scaled score' **range of 1 - 36 (inclusive)**. The composite score is calculated by averaging the 4 sections' scores, thus giving a possible **composite score range of 1 - 36**.

A quick visual inspection of the act_2017 CSV file indicates that one of the values in the dataset (in column 'Composite') has an incorrect 'x' attached to the numerical value, thus changing the data type of the entire column to *object*. This will have to be corrected, and all other values will also have to be checked using code, to ensure that they are within the correct ranges and of the correct data types (which may have been missed during the visual inspection).

In [15]:
# Checking the dataset info to verify data types of the columns and any missing values.

act17.info()

# Column 'Composite' has incorrect data type 'object' when it should be 'float'.
# This is because of the incorrect 'x' attached to one of the numerical values
# identified during the visual inspection above.

# No missing values are present in the dataset.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
State            52 non-null object
Participation    52 non-null object
English          52 non-null float64
Math             52 non-null float64
Reading          52 non-null float64
Science          52 non-null float64
Composite        52 non-null object
dtypes: float64(4), object(3)
memory usage: 2.9+ KB


In [16]:
# Checking the dataset info to verify data types of the columns and any missing values.

sat17.info()

# All numerical columns in this dataset have the correct data type int.
# No missing values are present in the dataset.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null object
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


#### 4c. Fix any errors you identified

**The data is available** so there's no need to guess or calculate anything. If you didn't find any errors, continue to the next step.

In [17]:
# Checking min and max values in the SAT 2017 dataset columns to ensure they are within the correct ranges.

# Correct ranges as stated above in section 4b.
# EBRW and Math sections' scores: 200 - 800 (inclusive)
# Total score: 400 - 1600

sat17.agg(['min', 'max'])

State Participation  Evidence-Based Reading and Writing  Math  Total
min  Alabama           10%                                 482    52    950
max  Wyoming           96%                                 644   651   1295

In [18]:
# Above table shows min value in Math column 52 < 200.
# All other min & max numerical values are within the correct ranges.

# Identifying the rows with Math score < 200.

sat17.loc[sat17['Math']<200]

State Participation  Evidence-Based Reading and Writing  Math  Total
20  Maryland           69%                                 536    52   1060

In [19]:
# Cross-referencing with the source, correct value for Math score for MAryland is 524.

# Replacing 52 in the above row with 524.

sat17.loc[sat17['State']=="Maryland", 'Math'] = 524

In [20]:
# Checking again to ensure the value is correctly replaced.

sat17.loc[sat17['State']=="Maryland"]

State Participation  Evidence-Based Reading and Writing  Math  Total
20  Maryland           69%                                 536   524   1060

In [21]:
# Checking min and max values in the ACT 2017 dataset columns to ensure they are within the correct ranges.

# Correct ranges as stated above in section 4b.
# English, Math, Reading and Science sections' scores: 1 - 36 (inclusive)
# Composite score: 1 - 36 (may not be correct, will have to be checked again after chainging column to float data type)

act17.agg(['min', 'max'])

State Participation  English  Math  Reading  Science Composite
min  Alabama           60%     16.3  18.0     18.1      2.3      17.8
max  Wyoming           98%     25.5  25.3     26.0     24.9      25.5

#### 5. What are your data types? 
Display the data types of each feature. 

In [22]:
# Checking the data types of the columns.

sat17.dtypes

State                                 object
Participation                         object
Evidence-Based Reading and Writing     int64
Math                                   int64
Total                                  int64
dtype: object

In [23]:
# Checking the data types of the columns.

act17.dtypes

State             object
Participation     object
English          float64
Math             float64
Reading          float64
Science          float64
Composite         object
dtype: object

What did you learn?
- Do any of them seem odd?  
- Which ones are not as they should be?  

**Answer:** As explained in section 4b. above, the 'Composite' column in ACT 2017 dataset has an incorrect data type *object* (because of an incorrect 'x' attached to one of the numerical values). This needs to be converted to *float*.

Also, as explained in section 3. above, the 'Participation' columns in both datasets are of type *object* because of the presence of a % symbol after the numerical values. These should be converted to type *float* as well.

#### 6. Fix Incorrect Data Types
Based on what you discovered above, use appropriate methods to re-type incorrectly typed data.
- Define a function that will allow you to convert participation rates to an appropriate numeric type. Use `map` or `apply` to change these columns in each dataframe.

In [24]:
# For SAT 2017 dataset
# Converting 'Participation' column from type object to type float by removing the % symbol after all numerical values.
# lambda function replaces the '%' in the strings with '' and then converts them to float.
# All float values are divided by 100 because they are percentages.
# .map funtion maps the lambda function to all values in 'Participation' column.

sat17['Participation'] = sat17["Participation"].map(lambda pct : float(pct.replace("%", ""))/100)

In [25]:
# Checking the data type of the column to ensure it is now of type float.

sat17['Participation'].dtypes

dtype('float64')

In [26]:
# Repeating above steps for ACT 2017 dataset.

act17['Participation'] = act17["Participation"].map(lambda pct : float(pct.replace("%", ""))/100)

In [27]:
# Checking the data type of the column to ensure it is now of type float.

act17['Participation'].dtypes

dtype('float64')

- Fix any individual values preventing other columns from being the appropriate type.

In [28]:
# For ACT 2017 dataset
# Checking all unique values of 'Composite' column to look for any abnormalities.

act17['Composite'].unique()

array(['21.0', '19.2', '19.8', '19.7', '19.4', '22.8', '20.8', '25.2',
       '24.1', '24.2', '21.4', '19.0', '22.3', '22.6', '21.9', '21.7',
       '20.0', '19.5', '24.3', '23.6', '25.4', '21.5', '18.6', '20.4',
       '20.3', '17.8', '25.5', '23.9', '19.1', '22.0', '21.8', '23.7',
       '24.0', '18.7', '20.7', '23.8', '20.5', '20.2x'], dtype=object)

In [29]:
# Filtering out the row with the incorrect 'Composite' value.

act17[act17['Composite']=='20.2x']

State  Participation  English  Math  Reading  Science Composite
51  Wyoming            1.0     19.4  19.8     20.8     20.6     20.2x

In [30]:
# Replacing the incorrect value with the correct one.

act17.loc[act17['Composite']=='20.2x', 'Composite']='20.2'

# Checking again to ensure the value is correctly replaced.

act17[act17['State']=='Wyoming']

State  Participation  English  Math  Reading  Science Composite
51  Wyoming            1.0     19.4  19.8     20.8     20.6      20.2

- Finish your data modifications by making sure the columns are now typed appropriately.

In [31]:
act17['Composite'].dtypes

dtype('O')

In [32]:
# Converting 'Composite' column values to type float

act17['Composite'] = act17['Composite'].astype(float)

In [33]:
# Checking the data type of the column to ensure it is now of type float.

act17['Composite'].dtypes

dtype('float64')

In [34]:
# Repeating step from section 4c. above after changing 'Composite' to type float.
# Checking min and max values in the ACT 2017 dataset columns to ensure they are within the correct ranges.

# Correct ranges as stated above in section 4b.
# English, Math, Reading and Science sections' scores: 1 - 36 (inclusive)
# Composite score: 1 - 36

act17.agg(['min', 'max'])

State  Participation  English  Math  Reading  Science  Composite
min  Alabama           0.08     16.3  18.0     18.1      2.3       17.8
max  Wyoming           1.00     25.5  25.3     26.0     24.9       25.5

- Display the data types again to confirm they are correct.

In [35]:
sat17.head()

State  Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama           0.05                                 593   572   1165
1      Alaska           0.38                                 547   533   1080
2     Arizona           0.30                                 563   553   1116
3    Arkansas           0.03                                 614   594   1208
4  California           0.53                                 531   524   1055

In [36]:
sat17.dtypes

State                                  object
Participation                         float64
Evidence-Based Reading and Writing      int64
Math                                    int64
Total                                   int64
dtype: object

In [37]:
act17.head()

State  Participation  English  Math  Reading  Science  Composite
0  National           0.60     20.3  20.7     21.4     21.0       21.0
1   Alabama           1.00     18.9  18.4     19.7     19.4       19.2
2    Alaska           0.65     18.7  19.8     20.4     19.9       19.8
3   Arizona           0.62     18.6  19.8     20.1     19.8       19.7
4  Arkansas           1.00     18.9  19.0     19.7     19.5       19.4

In [38]:
act17.dtypes

State             object
Participation    float64
English          float64
Math             float64
Reading          float64
Science          float64
Composite        float64
dtype: object

#### 7. Rename Columns
Change the names of the columns to more expressive names so that you can tell the difference the SAT columns and the ACT columns. Your solution should map all column names being changed at once (no repeated singular name-changes). **We will be combining these data with some of the data from 2018, and so you should name columns in an appropriate way**.

**Guidelines**:
- Column names should be all lowercase (you will thank yourself when you start pushing data to SQL later in the course)
- Column names should not contain spaces (underscores will suffice--this allows for using the `df.column_name` method to access columns in addition to `df['column_name']`.
- Column names should be unique and informative (the only feature that we actually share between dataframes is the state).

In [39]:
# Defining a dict of new column names for SAT 2017 dataset.

sat17_new_columns_dict = {'State' : 'states',
                          'Participation' : 'sat17_participation',
                          'Evidence-Based Reading and Writing' : 'sat17_ebrw',
                          'Math' : 'sat17_math',
                          'Total' : 'sat17_total'                        
                         }

# Renaming column names permanently in the dataset.

sat17.rename(columns=sat17_new_columns_dict, inplace=True)

In [40]:
sat17.head()

states  sat17_participation  sat17_ebrw  sat17_math  sat17_total
0     Alabama                 0.05         593         572         1165
1      Alaska                 0.38         547         533         1080
2     Arizona                 0.30         563         553         1116
3    Arkansas                 0.03         614         594         1208
4  California                 0.53         531         524         1055

In [41]:
# Defining a dict of new column names for ACT 2017 dataset.

act17_new_columns_dict = {'State' : 'states',
                          'Participation' : 'act17_participation',
                          'English' : 'act17_english',
                          'Math' : 'act17_math',
                          'Reading' : 'act17_reading',
                          'Science' : 'act17_science',
                          'Composite' : 'act17_composite'                        
                         }

# Renaming column names permanently in the dataset.

act17.rename(columns=act17_new_columns_dict, inplace=True)

In [42]:
act17.head()

states  act17_participation  act17_english  act17_math  act17_reading  \
0  National                 0.60           20.3        20.7           21.4   
1   Alabama                 1.00           18.9        18.4           19.7   
2    Alaska                 0.65           18.7        19.8           20.4   
3   Arizona                 0.62           18.6        19.8           20.1   
4  Arkansas                 1.00           18.9        19.0           19.7   

   act17_science  act17_composite  
0           21.0             21.0  
1           19.4             19.2  
2           19.9             19.8  
3           19.8             19.7  
4           19.5             19.4

#### 8. Create a data dictionary

Now that we've fixed our data, and given it appropriate names, let's create a [data dictionary](http://library.ucmerced.edu/node/10249). 

A data dictionary provides a quick overview of features/variables/columns, alongside data types and descriptions. The more descriptive you can be, the more useful this document is.

Example of a Fictional Data Dictionary Entry: 

|Feature|Type|Dataset|Description|
|---|---|---|---|
|**county_pop**|*integer*|2010 census|The population of the county (units in thousands, where 2.5 represents 2500 people).| 
|**per_poverty**|*float*|2010 census|The percent of the county over the age of 18 living below the 200% of official US poverty rate (units percent to two decimal places 98.10 means 98.1%)|

[Here's a quick link to a short guide for formatting markdown in Jupyter notebooks](https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html).

Provided is the skeleton for formatting a markdown table, with columns headers that will help you create a data dictionary to quickly summarize your data, as well as some examples. **This would be a great thing to copy and paste into your custom README for this project.**

### Data dictionary


|Feature|Type|Dataset|Description|
|---|---|---|---| 
|**states**|object|ACT 2017, SAT 2017|The states (50) and federal districts (1) of the USA. May also contain one additional 'National' row of data.|
|**sat17_participation**|float|SAT 2017|The percentage of students from the graduating class of 2017 who took the 2017 SAT (units percent, where 0.02 represents 2%).|
|**sat17_ebrw**|int|SAT 2017|Average scaled scores for the 2017 SAT Evidence-Based Reading and Writing (EBRW) section (without decimal places, in a range of 200 - 800 (inclusive)).|
|**sat17_math**|int|SAT 2017|Average scaled scores for the 2017 SAT Math section (without decimal places, in a range of 200 - 800 (inclusive)).|
|**sat17_total**|int|SAT 2017|Average scaled total scores for the 2017 SAT (without decimal places, in a range of 400 - 1600 (inclusive)). In SAT, the *individual* total scores are calculated by adding the two sections' scores.|
|**act17_participation**|float|ACT 2017|The percentage of students from the graduating class of 2017 who took the 2017 ACT (units percent, where 0.02 represents 2%).|
|**act17_english**|float|ACT 2017|Average scaled scores for the 2017 ACT English section (with one decimal place, in a range of 1 - 36 (inclusive)).|
|**act17_math**|float|ACT 2017|Average scaled scores for the 2017 ACT Math section (with one decimal place, in a range of 1 - 36 (inclusive)).|
|**act17_reading**|float|ACT 2017|Average scaled scores for the 2017 ACT Reading section (with one decimal place, in a range of 1 - 36 (inclusive)).|
|**act17_science**|float|ACT 2017|Average scaled scores for the 2017 ACT Science section (with one decimal place, in a range of 1 - 36 (inclusive)).|
|**act17_composite**|float|ACT 2017|Average scaled composite scores for the 2017 ACT (with one decimal place, in a range of 1 - 36 (inclusive)). In ACT, the *individual* composite scores are calculated by averaging the four sections' scores.|

#### 9. Drop unnecessary rows

One of our dataframes contains an extra row. Identify and remove this from the dataframe.

In [43]:
# Identifying and filtering the extra row.

act17.iloc[:1]

states  act17_participation  act17_english  act17_math  act17_reading  \
0  National                  0.6           20.3        20.7           21.4   

   act17_science  act17_composite  
0           21.0             21.0

In [44]:
# Dropping the identified row permanently.

act17.drop(0, axis=0, inplace=True)

# Reseting the index to start from 0.

act17.reset_index(drop=True, inplace=True)

In [45]:
act17.head()

states  act17_participation  act17_english  act17_math  act17_reading  \
0     Alabama                 1.00           18.9        18.4           19.7   
1      Alaska                 0.65           18.7        19.8           20.4   
2     Arizona                 0.62           18.6        19.8           20.1   
3    Arkansas                 1.00           18.9        19.0           19.7   
4  California                 0.31           22.5        22.7           23.1   

   act17_science  act17_composite  
0           19.4             19.2  
1           19.9             19.8  
2           19.8             19.7  
3           19.5             19.4  
4           22.2             22.8

#### 10. Merge Dataframes

Join the 2017 ACT and SAT dataframes using the state in each dataframe as the key. Assign this to a new variable.

In [46]:
sat17.shape

(51, 5)

In [47]:
act17.shape

(51, 7)

In [48]:
# Merging the two dataframes on the 'states' column and storing it as a new dataframe 'sat_act_17'.

# We use 'outer' join to ensure no rows of data are lost in both the datasets.
# This will ensure no data loss in case the two datasets happen to have one (or more) state names
# which have not been typed identically in the two datasets by mistake (or for some other reason).

sat_act_17 = pd.merge(sat17, act17, on='states', how='outer')

In [49]:
# Checking shape of combined dataframe to ensure it still only has 51 rows of data.
# If combined dataframe has 51 rows of data, it means that all 51 state names matched together identically between the
# two datasets, and no empty values exist in the combined dataframe.

sat_act_17.shape

(51, 11)

In [50]:
sat_act_17.head()

states  sat17_participation  sat17_ebrw  sat17_math  sat17_total  \
0     Alabama                 0.05         593         572         1165   
1      Alaska                 0.38         547         533         1080   
2     Arizona                 0.30         563         553         1116   
3    Arkansas                 0.03         614         594         1208   
4  California                 0.53         531         524         1055   

   act17_participation  act17_english  act17_math  act17_reading  \
0                 1.00           18.9        18.4           19.7   
1                 0.65           18.7        19.8           20.4   
2                 0.62           18.6        19.8           20.1   
3                 1.00           18.9        19.0           19.7   
4                 0.31           22.5        22.7           23.1   

   act17_science  act17_composite  
0           19.4             19.2  
1           19.9             19.8  
2           19.8             19.7  
3           19.5             19.4  
4           22.2             22.8

#### 11. Save your cleaned, merged dataframe

Use a relative path to save out your data as `combined_2017.csv`.

In [51]:
# Exporting the merged dataframe to a CSV file using a relative path.

sat_act_17.to_csv('../data/combined_2017.csv', index=False)

## 2018 Data Import and Cleaning

Links to the 2018 ACT and SAT data are provided in the README. These data live in PDFs, and so you'll get to enjoy practicing some *manual* data collection. Save these data as a CSV in your `data` directory, and import, explore, and clean these data in the same way you did above. **Make sure you comment on your steps so it is clear *why* you are doing each process**.

In [52]:
# Reading the CSV files using relative filepaths

sat18 = pd.read_csv("../data/sat_2018.csv")
act18 = pd.read_csv("../data/act_2018_updated.csv")

In [53]:
# Printing a few lines of each dataframe to ensure that it was read properly.

sat18.head(10)

State Participation  Evidence-Based Reading and Writing  \
0               Alabama            6%                                 595   
1                Alaska           43%                                 562   
2               Arizona           29%                                 577   
3              Arkansas            5%                                 592   
4            California           60%                                 540   
5              Colorado          100%                                 519   
6           Connecticut          100%                                 535   
7              Delaware          100%                                 505   
8  District of Columbia           92%                                 497   
9               Florida           56%                                 550   

   Math  Total  
0   571   1166  
1   544   1106  
2   572   1149  
3   576   1169  
4   536   1076  
5   506   1025  
6   519   1053  
7   492    998  
8   480    977  
9   549   1099

In [54]:
sat18.tail()

State Participation  Evidence-Based Reading and Writing  Math  \
46       Virginia           68%                                 567   550   
47     Washington           69%                                 543   538   
48  West Virginia           28%                                 513   486   
49      Wisconsin            3%                                 641   653   
50        Wyoming            3%                                 633   625   

    Total  
46   1117  
47   1081  
48    999  
49   1294  
50   1257

In [55]:
# Checking the number of rows in the dataframe to ensure it is 51 (same as sat17).

len(sat18)

51

In [56]:
# Checking the number of values in the 'State' column to ensure it is 51 (same as sat17).

len(sat18['State'])

51

In [57]:
# Checking the unique values in the 'State' column.

sat18['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'District of Columbia',
       'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Oklahoma', 'Ohio', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

In [58]:
act18.head(10)

State  Percentage of Students Tested  \
0               Alabama                            100   
1                Alaska                             33   
2               Arizona                             66   
3              Arkansas                            100   
4            California                             27   
5              Colorado                             30   
6           Connecticut                             26   
7              Delaware                             17   
8  District of Columbia                             32   
9               Florida                             66   

   Average Composite Score  Average English Score  Average Math Score  \
0                     19.1                   18.9                18.3   
1                     20.8                   19.8                20.6   
2                     19.2                   18.2                19.4   
3                     19.4                   19.1                18.9   
4                     22.7                   22.5                22.5   
5                     23.9                   23.9                23.2   
6                     25.6                   26.0                24.8   
7                     23.2                   23.7                23.1   
8                     23.6                   23.7                22.7   
9                     19.9                   19.2                19.3   

   Average Reading Score  Average Science Score  
0                   19.6                   19.0  
1                   21.6                   20.7  
2                   19.5                   19.2  
3                   19.7                   19.4  
4                   23.0                   22.1  
5                   24.4                   23.5  
6                   26.1                   24.9  
7                   24.5                   23.4  
8                   24.4                   23.0  
9                   21.1                   19.5

In [59]:
act18.tail()

State  Percentage of Students Tested  Average Composite Score  \
46       Virginia                             24                     23.9   
47     Washington                             24                     22.2   
48  West Virginia                             65                     20.3   
49      Wisconsin                            100                     20.5   
50        Wyoming                            100                     20.0   

    Average English Score  Average Math Score  Average Reading Score  \
46                   23.8                23.3                   24.7   
47                   21.4                22.2                   22.7   
48                   19.8                19.4                   21.3   
49                   19.8                20.3                   20.6   
50                   19.0                19.7                   20.6   

    Average Science Score  
46                   23.5  
47                   22.0  
48                   20.4  
49                   20.8  
50                   20.3

In [60]:
# Checking the number of rows in the dataframe to ensure it is 51 (same as act17).

len(act18)

51

In [61]:
# Checking the number of values in the 'State' column to ensure it is 51 (same as act17).

len(act18['State'])

51

In [62]:
# Checking the unique values in the 'State' column.

act18['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'District of Columbia',
       'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

**Description of SAT 2018 and ACT 2018 datasets**

**SAT 2018:** Similar to SAT 2017, this dataset *(imported here as 'sat18')* provides a breakdown of the 2018 SAT participation rates *(column 'Participation')*, average scores for EBRW and Math sections *(columns 'Evidence-Based Reading and Writing' and 'Math', respectively)* and the average aggregate scores *(column 'Total')* for all 50 states and 1 federal district (District of Columbia) of the USA *(column 'State', arranged alphabetically)*.

The values in the 'State' column are *strings*. The numerical values of the scores in the dataset do not carry any decimal places, and so should be expected to be *integers*. The values in the 'Participation' column carry a numerical value followed by a % symbol. So, these should be expected to be *strings* which would need to be converted to *floats*.

**ACT 2018:** Similar to ACT 2017, this dataset *(imported here as 'act18')* provides a breakdown of the 2018 ACT participation rates *(column 'Percentage of Students Tested')*, average scores for English, Math, Reading and Science sections *(columns 'Average English Score', 'Average Math Score', 'Average Reading Score' and 'Average Science Score', respectively)* and the average composite scores *(column 'Average Composite Score')* for all 50 states and 1 federal district (District of Columbia) of the USA *(column 'State', arranged alphabetically)*.

The values in the 'State' column are *strings*. The numerical values of the scores in the dataset carry decimal places, and so should be expected to be *floats*. The values in the 'Percentage of Students Tested' column carry a numerical value, so should be expected to be *integers*. However, in order to maintain consistency with SAT 2017, ACT 2017 & SAT 2018 datasets, these values would need to be converted to *floats*.

**Quick visual inspection of the datasets**

*The description in this section is ONLY based on a quick visual inspection of the CSV files. This method may be usable here because the datasets are relatively small in size (only 5-7 columns, ~51 rows).*

As explained in section 4b. above, for the **SAT**, the ranges for possible 'scaled scores' for each of the **2 setions** are **200 - 800 (inclusive)**. The total score is calculated by adding the 2 sections' scores, thus giving a possible **total score range of 400 - 1600**.

A quick visual inspection of the sat_2018 CSV file did not reveal any abnormal values. All values will still have to be checked using code, to ensure that they are within the correct ranges and of the correct data types (which may have been missed during the visual inspection).

As explained in section 4b. above, for the **ACT**, the ranges for possible 'scaled scores' for each of the **4 sections** are **1 - 36 (inclusive)**. The composite score is calculated by averaging the 4 sections' scores, thus giving a possible **composite score range of 1 - 36**.

A quick visual inspection of the act_2018_updated CSV file did not reveal any abnormal values. All values will still have to be checked using code, to ensure that they are within the correct ranges and of the correct data types (which may have been missed during the visual inspection).

In [63]:
# Checking the dataset info to verify data types of the columns and any missing values.

sat18.info()

# All numerical columns in this dataset have the correct data type int.
# No missing values are present in the dataset.
# 'Participation' column will have to be converted to type 'float' (similar to SAT 2017 and ACT 2017 datasets).

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 5 columns):
State                                 51 non-null object
Participation                         51 non-null object
Evidence-Based Reading and Writing    51 non-null int64
Math                                  51 non-null int64
Total                                 51 non-null int64
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


In [64]:
# Checking the dataset info to verify data types of the columns and any missing values.

act18.info()

# All numerical columns in this dataset have the correct data type float.
# No missing values are present in the dataset.
# 'Participation' column will have to be converted to type 'float' (similar to SAT 2017 and ACT 2017 datasets).

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 7 columns):
State                            51 non-null object
Percentage of Students Tested    51 non-null int64
Average Composite Score          51 non-null float64
Average English Score            51 non-null float64
Average Math Score               51 non-null float64
Average Reading Score            51 non-null float64
Average Science Score            51 non-null float64
dtypes: float64(5), int64(1), object(1)
memory usage: 2.9+ KB


In [65]:
# Checking min and max values in the SAT 2018 dataset columns to ensure they are within the correct ranges.

# Correct ranges as stated above in section 4b.
# EBRW and Math sections' scores: 200 - 800 (inclusive)
# Total score: 400 - 1600

sat18.agg(['min', 'max'])

State Participation  Evidence-Based Reading and Writing  Math  Total
min  Alabama           10%                                 480   480    977
max  Wyoming           99%                                 643   655   1298

In [66]:
# Checking min and max values in the ACT 2018 dataset columns to ensure they are within the correct ranges.

# Correct ranges as stated above in section 4b.
# English, Math, Reading and Science sections' scores: 1 - 36 (inclusive)
# Composite score: 1 - 36

act18.agg(['min', 'max'])

State  Percentage of Students Tested  Average Composite Score  \
min  Alabama                              7                     17.7   
max  Wyoming                            100                     25.6   

     Average English Score  Average Math Score  Average Reading Score  \
min                   16.6                17.8                   18.0   
max                   26.0                25.2                   26.1   

     Average Science Score  
min                   17.9  
max                   24.9

In [67]:
# For SAT 2018 dataset
# Converting 'Participation' column from type object to type float by removing the % symbol after all numerical values.
# lambda function replaces the '%' in the strings with '' and then converts them to float.
# All float values are divided by 100 because they are percentages.
# .map funtion maps the lambda function to all values in 'Participation' column.

sat18['Participation'] = sat18["Participation"].map(lambda pct : float(pct.replace("%", ""))/100)

In [68]:
# Checking the data type of the columns again to ensure they are now all correct.

sat18.dtypes

State                                  object
Participation                         float64
Evidence-Based Reading and Writing      int64
Math                                    int64
Total                                   int64
dtype: object

In [69]:
sat18.head()

State  Participation  Evidence-Based Reading and Writing  Math  Total
0     Alabama           0.06                                 595   571   1166
1      Alaska           0.43                                 562   544   1106
2     Arizona           0.29                                 577   572   1149
3    Arkansas           0.05                                 592   576   1169
4  California           0.60                                 540   536   1076

In [70]:
# For ACT 2018 dataset
# Converting 'Percentage of Students Tested' column from type int to type float.
# All float values are divided by 100 because they are percentages.

act18['Percentage of Students Tested'] = act18["Percentage of Students Tested"].astype(float) / 100

In [71]:
# Checking the data type of the columns again to ensure they are now all correct.

act18.dtypes

State                             object
Percentage of Students Tested    float64
Average Composite Score          float64
Average English Score            float64
Average Math Score               float64
Average Reading Score            float64
Average Science Score            float64
dtype: object

In [72]:
act18.head()

State  Percentage of Students Tested  Average Composite Score  \
0     Alabama                           1.00                     19.1   
1      Alaska                           0.33                     20.8   
2     Arizona                           0.66                     19.2   
3    Arkansas                           1.00                     19.4   
4  California                           0.27                     22.7   

   Average English Score  Average Math Score  Average Reading Score  \
0                   18.9                18.3                   19.6   
1                   19.8                20.6                   21.6   
2                   18.2                19.4                   19.5   
3                   19.1                18.9                   19.7   
4                   22.5                22.5                   23.0   

   Average Science Score  
0                   19.0  
1                   20.7  
2                   19.2  
3                   19.4  
4                   22.1

In [73]:
# Defining a dict of new column names for SAT 2018 dataset.

sat18_new_columns_dict = {'State' : 'states',
                          'Participation' : 'sat18_participation',
                          'Evidence-Based Reading and Writing' : 'sat18_ebrw',
                          'Math' : 'sat18_math',
                          'Total' : 'sat18_total'                        
                         }

# Renaming column names permanently in the dataset.

sat18.rename(columns=sat18_new_columns_dict, inplace=True)

In [74]:
sat18.head()

states  sat18_participation  sat18_ebrw  sat18_math  sat18_total
0     Alabama                 0.06         595         571         1166
1      Alaska                 0.43         562         544         1106
2     Arizona                 0.29         577         572         1149
3    Arkansas                 0.05         592         576         1169
4  California                 0.60         540         536         1076

In [75]:
# Defining a dict of new column names for ACT 2018 dataset.

act18_new_columns_dict = {'State' : 'states',
                          'Percentage of Students Tested' : 'act18_participation',
                          'Average English Score' : 'act18_english',
                          'Average Math Score' : 'act18_math',
                          'Average Reading Score' : 'act18_reading',
                          'Average Science Score' : 'act18_science',
                          'Average Composite Score' : 'act18_composite'                        
                         }

# Renaming column names permanently in the dataset.

act18.rename(columns=act18_new_columns_dict, inplace=True)

In [76]:
act18.head()

states  act18_participation  act18_composite  act18_english  \
0     Alabama                 1.00             19.1           18.9   
1      Alaska                 0.33             20.8           19.8   
2     Arizona                 0.66             19.2           18.2   
3    Arkansas                 1.00             19.4           19.1   
4  California                 0.27             22.7           22.5   

   act18_math  act18_reading  act18_science  
0        18.3           19.6           19.0  
1        20.6           21.6           20.7  
2        19.4           19.5           19.2  
3        18.9           19.7           19.4  
4        22.5           23.0           22.1

In [77]:
# Re-arranging the order of columns to maintain consistency with ACT 2017 dataset.

act18 = act18[['states', 'act18_participation', 'act18_english', 'act18_math', 'act18_reading', 'act18_science', 'act18_composite']]

In [78]:
act18.head()

states  act18_participation  act18_english  act18_math  act18_reading  \
0     Alabama                 1.00           18.9        18.3           19.6   
1      Alaska                 0.33           19.8        20.6           21.6   
2     Arizona                 0.66           18.2        19.4           19.5   
3    Arkansas                 1.00           19.1        18.9           19.7   
4  California                 0.27           22.5        22.5           23.0   

   act18_science  act18_composite  
0           19.0             19.1  
1           20.7             20.8  
2           19.2             19.2  
3           19.4             19.4  
4           22.1             22.7

### Data dictionary


|Feature|Type|Dataset|Description|
|---|---|---|---| 
|**states**|object|ACT 2018, SAT 2018|The states (50) and federal districts (1) of the USA.|
|**sat18_participation**|float|SAT 2018|The percentage of students from the graduating class of 2018 who took the 2018 SAT (units percent, where 0.02 represents 2%).|
|**sat18_ebrw**|int|SAT 2018|Average scaled scores for the 2018 SAT Evidence-Based Reading and Writing (EBRW) section (without decimal places, in a range of 200 - 800 (inclusive)).|
|**sat18_math**|int|SAT 2018|Average scaled scores for the 2018 SAT Math section (without decimal places, in a range of 200 - 800 (inclusive)).|
|**sat18_total**|int|SAT 2018|Average scaled total scores for the 2018 SAT (without decimal places, in a range of 400 - 1600 (inclusive)). In SAT, the *individual* total scores are calculated by adding the two sections' scores.|
|**act18_participation**|float|ACT 2018|The percentage of students from the graduating class of 2018 who took the 2018 ACT (units percent, where 0.02 represents 2%).|
|**act18_english**|float|ACT 2018|Average scaled scores for the 2018 ACT English section (with one decimal place, in a range of 1 - 36 (inclusive)).|
|**act18_math**|float|ACT 2018|Average scaled scores for the 2018 ACT Math section (with one decimal place, in a range of 1 - 36 (inclusive)).|
|**act18_reading**|float|ACT 2018|Average scaled scores for the 2018 ACT Reading section (with one decimal place, in a range of 1 - 36 (inclusive)).|
|**act18_science**|float|ACT 2018|Average scaled scores for the 2018 ACT Science section (with one decimal place, in a range of 1 - 36 (inclusive)).|
|**act18_composite**|float|ACT 2018|Average scaled composite scores for the 2018 ACT (with one decimal place, in a range of 1 - 36 (inclusive)). In ACT, the *individual* composite scores are calculated by averaging the four sections' scores.|

In [79]:
sat18.shape

(51, 5)

In [80]:
act18.shape

(51, 7)

In [81]:
# Merging the two dataframes on the 'states' column and storing it as a new dataframe 'sat_act_18'.

# We use 'outer' join to ensure no rows of data are lost in both the datasets.
# This will ensure no data loss in case the two datasets happen to have one (or more) state names
# which have not been typed identically in the two datasets by mistake (or for some other reason).

sat_act_18 = pd.merge(sat18, act18, on='states', how='outer')

In [82]:
# Checking shape of combined dataframe to ensure it still only has 51 rows of data.
# If combined dataframe has 51 rows of data, it means that all 51 state names matched together identically between the
# two datasets, and no empty values exist in the combined dataframe.

sat_act_18.shape

(51, 11)

In [83]:
sat_act_18.head()

states  sat18_participation  sat18_ebrw  sat18_math  sat18_total  \
0     Alabama                 0.06         595         571         1166   
1      Alaska                 0.43         562         544         1106   
2     Arizona                 0.29         577         572         1149   
3    Arkansas                 0.05         592         576         1169   
4  California                 0.60         540         536         1076   

   act18_participation  act18_english  act18_math  act18_reading  \
0                 1.00           18.9        18.3           19.6   
1                 0.33           19.8        20.6           21.6   
2                 0.66           18.2        19.4           19.5   
3                 1.00           19.1        18.9           19.7   
4                 0.27           22.5        22.5           23.0   

   act18_science  act18_composite  
0           19.0             19.1  
1           20.7             20.8  
2           19.2             19.2  
3           19.4             19.4  
4           22.1             22.7

In [84]:
# Exporting the merged dataframe to a CSV file using a relative path.

sat_act_18.to_csv('../data/combined_2018.csv', index=False)

#### Combine your 2017 and 2018 data into a single dataframe
Joining on state names should work, assuming you formatted all your state names identically. Make sure none of your columns (other than state) have identical names. Do yourself a favor and decide if you're encoding participation rates as floats or integers and standardize this across your datasets.

Save the contents of this merged dataframe as `final.csv`.

**Use this combined dataframe for the remainder of the project**.

In [85]:
# Merging the two dataframes on the 'states' column and storing it as a new dataframe 'satact_1718'.

# We use 'outer' join to ensure no rows of data are lost in both the datasets.
# This will ensure no data loss in case the two datasets happen to have one (or more) state names
# which have not been typed identically in the two datasets by mistake (or for some other reason).

satact_1718 = pd.merge(sat_act_17, sat_act_18, on='states', how='outer')

In [86]:
satact_1718.head()

states  sat17_participation  sat17_ebrw  sat17_math  sat17_total  \
0     Alabama                 0.05         593         572         1165   
1      Alaska                 0.38         547         533         1080   
2     Arizona                 0.30         563         553         1116   
3    Arkansas                 0.03         614         594         1208   
4  California                 0.53         531         524         1055   

   act17_participation  act17_english  act17_math  act17_reading  \
0                 1.00           18.9        18.4           19.7   
1                 0.65           18.7        19.8           20.4   
2                 0.62           18.6        19.8           20.1   
3                 1.00           18.9        19.0           19.7   
4                 0.31           22.5        22.7           23.1   

   act17_science  ...  sat18_participation  sat18_ebrw  sat18_math  \
0           19.4  ...                 0.06         595         571   
1           19.9  ...                 0.43         562         544   
2           19.8  ...                 0.29         577         572   
3           19.5  ...                 0.05         592         576   
4           22.2  ...                 0.60         540         536   

   sat18_total  act18_participation  act18_english  act18_math  act18_reading  \
0         1166                 1.00           18.9        18.3           19.6   
1         1106                 0.33           19.8        20.6           21.6   
2         1149                 0.66           18.2        19.4           19.5   
3         1169                 1.00           19.1        18.9           19.7   
4         1076                 0.27           22.5        22.5           23.0   

   act18_science  act18_composite  
0           19.0             19.1  
1           20.7             20.8  
2           19.2             19.2  
3           19.4             19.4  
4           22.1             22.7  

[5 rows x 21 columns]

In [87]:
# Exporting the merged dataframe to a CSV file using a relative path.

satact_1718.to_csv('../data/final.csv', index=False)

## Exploratory Data Analysis


### Summary Statistics
Transpose the output of pandas `describe` method to create a quick overview of each numeric feature.

In [89]:
satact_1718.describe().T

count         mean        std     min       25%      50%  \
sat17_participation   51.0     0.398039   0.352766    0.02     0.040     0.38   
sat17_ebrw            51.0   569.117647  45.666901  482.00   533.500   559.00   
sat17_math            51.0   556.882353  47.121395  468.00   523.500   548.00   
sat17_total           51.0  1126.098039  92.494812  950.00  1055.500  1107.00   
act17_participation   51.0     0.652549   0.321408    0.08     0.310     0.69   
act17_english         51.0    20.931373   2.353677   16.30    19.000    20.70   
act17_math            51.0    21.182353   1.981989   18.00    19.400    20.90   
act17_reading         51.0    22.013725   2.067271   18.10    20.450    21.80   
act17_science         51.0    21.041176   3.182463    2.30    19.900    21.30   
act17_composite       51.0    21.519608   2.020695   17.80    19.800    21.40   
sat18_participation   51.0     0.457451   0.373143    0.02     0.045     0.52   
sat18_ebrw            51.0   563.686275  47.502627  480.00   534.500   552.00   
sat18_math            51.0   556.235294  47.772623  480.00   522.500   544.00   
sat18_total           51.0  1120.019608  94.155083  977.00  1057.500  1098.00   
act18_participation   51.0     0.616471   0.340810    0.07     0.285     0.66   
act18_english         51.0    20.988235   2.446356   16.60    19.100    20.20   
act18_math            51.0    21.125490   2.035765   17.80    19.400    20.70   
act18_reading         51.0    22.015686   2.167245   18.00    20.450    21.60   
act18_science         51.0    21.345098   1.870114   17.90    19.850    21.10   
act18_composite       51.0    21.486275   2.106278   17.70    19.950    21.30   

                          75%     max  
sat17_participation     0.660     1.0  
sat17_ebrw            613.000   644.0  
sat17_math            599.000   651.0  
sat17_total          1212.000  1295.0  
act17_participation     1.000     1.0  
act17_english          23.300    25.5  
act17_math             23.100    25.3  
act17_reading          24.150    26.0  
act17_science          22.750    24.9  
act17_composite        23.600    25.5  
sat18_participation     0.775     1.0  
sat18_ebrw            610.500   643.0  
sat18_math            593.500   655.0  
sat18_total          1204.000  1298.0  
act18_participation     1.000     1.0  
act18_english          23.700    26.0  
act18_math             23.150    25.2  
act18_reading          24.100    26.1  
act18_science          23.050    24.9  
act18_composite        23.550    25.6

#### Manually calculate standard deviation

$$\sigma = \sqrt{\frac{1}{n}\sum_{i=1}^n(x_i - \mu)^2}$$

- Write a function to calculate standard deviation using the formula above

In [98]:
# Function to calculate standard deviation of a series of data.
# The function takes the column as an input and returns the standard deviation of the values in the column (calculated using the above formula).

def calculate_stdev(col):
    n = len(col)
    mean = col.sum() / n
    diff = 0
    for val in col:
        diff += (val-mean)**2
    std = (diff/n)**0.5
    return std

- Use a **dictionary comprehension** to apply your standard deviation function to each numeric column in the dataframe.  **No loops**  
- Assign the output to variable `sd` as a dictionary where: 
    - Each column name is now a key 
    - That standard deviation of the column is the value 
     
*Example Output :* `{'ACT_Math': 120, 'ACT_Reading': 120, ...}`

In [112]:
# Calculating stdev of all numeric columns and storing them in a new variable using dictionary comprehension.
# 1. Filter out the numeric columns using .iloc and convert the filtered dataframe into a dictionary using dict().
# 2. Using dictionary comprehension, store the column names as keys and the returned stdev values from calculate_stdev function
#    as values into a new dictionary called 'sd'.

sd = {col_name : calculate_stdev(vals) for col_name, vals in dict(satact_1718.iloc[:, 1:]).items()}

In [113]:
sd

{'sat17_participation': 0.3492907076664507,
 'sat17_ebrw': 45.21697020437866,
 'sat17_math': 46.65713364485503,
 'sat17_total': 91.58351056778743,
 'act17_participation': 0.3182417575123181,
 'act17_english': 2.3304876369363363,
 'act17_math': 1.9624620273436781,
 'act17_reading': 2.0469029314842646,
 'act17_science': 3.151107895464408,
 'act17_composite': 2.000786081581989,
 'sat18_participation': 0.3694661922353941,
 'sat18_ebrw': 47.03460978357609,
 'sat18_math': 47.30194550378352,
 'sat18_total': 93.22742384464433,
 'act18_participation': 0.33745194881997503,
 'act18_english': 2.4222536143202795,
 'act18_math': 2.0157072555557174,
 'act18_reading': 2.145891884510421,
 'act18_science': 1.851688548483354,
 'act18_composite': 2.0855261815801147}

Do your manually calculated standard deviations match up with the output from pandas `describe`? What about numpy's `std` method?

In [116]:
# Using np.std to calculate stdev of 'sat17_total' column.

np.std(satact_1718['sat17_total'])

91.58351056778743

In [120]:
# Using pd.DataFrame.std to calculate stdev of 'sat17_total' column.
# Argument delta degrees of freedom (ddof) set to 1 by default.

satact_1718['sat17_total'].std()

92.49481172519046

In [115]:
# Using pd.DataFrame.std to calculate stdev of 'sat17_total' column.
# Setting delta degrees of freedom (ddof) to 0.

satact_1718['sat17_total'].std(ddof=0)

91.58351056778743

**Answer:** We notice that the manually calculated values of standard deviation are same as the ones calculated using *np.std* method. This is because *np.std* has an argument for delta degrees of freedom *(ddof)*, which is set to 0 by default.

The values we obtained using *pd.DataFrame.describe()* were different because the argument *ddof* is set to 1 by default. This is also the case in *pd.DataFrame.std()*. When we explicitly set *ddof = 0*, we obtain the stdev values which are same as *np.std* and our manually calculating function *(calculate_stdev(col))*.

#### Investigate trends in the data
Using sorting and/or masking (along with the `.head` method to not print our entire dataframe), consider the following questions:

- Which states have the highest and lowest participation rates for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Which states have the highest and lowest mean total/composite scores for the:
    - 2017 SAT?
    - 2018 SAT?
    - 2017 ACT?
    - 2018 ACT?
- Do any states with 100% participation on a given test have a rate change year-to-year?
- Do any states show have >50% participation on *both* tests either year?

Based on what you've just observed, have you identified any states that you're especially interested in? **Make a note of these and state *why* you think they're interesting**.

**You should comment on your findings at each step in a markdown cell below your code block**. Make sure you include at least one example of sorting your dataframe by a column, and one example of using boolean filtering (i.e., masking) to select a subset of the dataframe.

In [122]:
satact_1718.head()

states  sat17_participation  sat17_ebrw  sat17_math  sat17_total  \
0     Alabama                 0.05         593         572         1165   
1      Alaska                 0.38         547         533         1080   
2     Arizona                 0.30         563         553         1116   
3    Arkansas                 0.03         614         594         1208   
4  California                 0.53         531         524         1055   

   act17_participation  act17_english  act17_math  act17_reading  \
0                 1.00           18.9        18.4           19.7   
1                 0.65           18.7        19.8           20.4   
2                 0.62           18.6        19.8           20.1   
3                 1.00           18.9        19.0           19.7   
4                 0.31           22.5        22.7           23.1   

   act17_science  ...  sat18_participation  sat18_ebrw  sat18_math  \
0           19.4  ...                 0.06         595         571   
1           19.9  ...                 0.43         562         544   
2           19.8  ...                 0.29         577         572   
3           19.5  ...                 0.05         592         576   
4           22.2  ...                 0.60         540         536   

   sat18_total  act18_participation  act18_english  act18_math  act18_reading  \
0         1166                 1.00           18.9        18.3           19.6   
1         1106                 0.33           19.8        20.6           21.6   
2         1149                 0.66           18.2        19.4           19.5   
3         1169                 1.00           19.1        18.9           19.7   
4         1076                 0.27           22.5        22.5           23.0   

   act18_science  act18_composite  
0           19.0             19.1  
1           20.7             20.8  
2           19.2             19.2  
3           19.4             19.4  
4           22.1             22.7  

[5 rows x 21 columns]

In [136]:
# Creating masks to obtain min & max values of 'sat17_participation' column
min_sat17_par = satact_1718['sat17_participation']==satact_1718['sat17_participation'].min()
max_sat17_par = satact_1718['sat17_participation']==satact_1718['sat17_participation'].max()

# Applying the masks to obtain states with the lowest OR highest participation rates,
# and then sorting the rows by participation rates in ascending order.
minmax_sat17_par = satact_1718[['states', 'sat17_participation']].loc[(min_sat17_par) | (max_sat17_par)].sort_values(by='sat17_participation')
minmax_sat17_par

states  sat17_participation
15                  Iowa                 0.02
24           Mississippi                 0.02
34          North Dakota                 0.02
6            Connecticut                 1.00
7               Delaware                 1.00
8   District of Columbia                 1.00
22              Michigan                 1.00

In [137]:
# Creating masks to obtain min & max values of 'sat18_participation' column
min_sat18_par = satact_1718['sat18_participation']==satact_1718['sat18_participation'].min()
max_sat18_par = satact_1718['sat18_participation']==satact_1718['sat18_participation'].max()

# Applying the masks to obtain states with the lowest OR highest participation rates,
# and then sorting the rows by participation rates in ascending order.
minmax_sat18_par = satact_1718[['states', 'sat18_participation']].loc[(min_sat18_par) | (max_sat18_par)].sort_values(by='sat18_participation')
minmax_sat18_par

states  sat18_participation
34  North Dakota                 0.02
5       Colorado                 1.00
6    Connecticut                 1.00
7       Delaware                 1.00
12         Idaho                 1.00
22      Michigan                 1.00

In [161]:
# Merging the above two dataframes on the 'states' column.
# We use 'inner' join to filter only the states that appear in both the above two dataframes.

sat1718_par_compare = pd.merge(minmax_sat17_par, minmax_sat18_par, on='states', how='inner')
sat1718_par_compare

states  sat17_participation  sat18_participation
0  North Dakota                 0.02                 0.02
1   Connecticut                 1.00                 1.00
2      Delaware                 1.00                 1.00
3      Michigan                 1.00                 1.00

From the above, we observe the following:

- **2017 SAT**
    - **Lowest participation rates (2%) :** 3 states (Iowa, Mississippi and North Dakota)
    - **Highest participation rates (100%) :** 4 states (Connecticut, Delaware, District of Columbia and Michigan)
- **2018 SAT**
    - **Lowest participation rates (2%) :** 1 state (North Dakota)
    - **Highest participation rates (100%) :** 5 states (Colorado, Connecticut, Delaware, Idaho and Michigan)
- North Dakota consistently has the lowest SAT participation rates of 2%.
- Connecticut, Delaware and Michigan consistenly have the highest SAT participation rates of 100%.

In [140]:
# Creating masks to obtain min & max values of 'act17_participation' column
min_act17_par = satact_1718['act17_participation']==satact_1718['act17_participation'].min()
max_act17_par = satact_1718['act17_participation']==satact_1718['act17_participation'].max()

# Applying the masks to obtain states with the lowest OR highest participation rates,
# and then sorting the rows by participation rates in ascending order.
minmax_act17_par = satact_1718[['states', 'act17_participation']].loc[(min_act17_par) | (max_act17_par)].sort_values(by='act17_participation')
minmax_act17_par

states  act17_participation
19           Maine                 0.08
0          Alabama                 1.00
44            Utah                 1.00
42       Tennessee                 1.00
40  South Carolina                 1.00
36        Oklahoma                 1.00
33  North Carolina                 1.00
28          Nevada                 1.00
26         Montana                 1.00
25        Missouri                 1.00
24     Mississippi                 1.00
23       Minnesota                 1.00
18       Louisiana                 1.00
17        Kentucky                 1.00
5         Colorado                 1.00
3         Arkansas                 1.00
49       Wisconsin                 1.00
50         Wyoming                 1.00

In [141]:
# Creating masks to obtain min & max values of 'act18_participation' column
min_act18_par = satact_1718['act18_participation']==satact_1718['act18_participation'].min()
max_act18_par = satact_1718['act18_participation']==satact_1718['act18_participation'].max()

# Applying the masks to obtain states with the lowest OR highest participation rates,
# and then sorting the rows by participation rates in ascending order.
minmax_act18_par = satact_1718[['states', 'act18_participation']].loc[(min_act18_par) | (max_act18_par)].sort_values(by='act18_participation')
minmax_act18_par

states  act18_participation
19           Maine                 0.07
0          Alabama                 1.00
44            Utah                 1.00
42       Tennessee                 1.00
40  South Carolina                 1.00
36        Oklahoma                 1.00
35            Ohio                 1.00
33  North Carolina                 1.00
28          Nevada                 1.00
27        Nebraska                 1.00
26         Montana                 1.00
25        Missouri                 1.00
24     Mississippi                 1.00
18       Louisiana                 1.00
17        Kentucky                 1.00
3         Arkansas                 1.00
49       Wisconsin                 1.00
50         Wyoming                 1.00

In [162]:
# Merging the above two dataframes on the 'states' column.
# We use 'inner' join to filter only the states that appear in both the above two dataframes.

act1718_par_compare = pd.merge(minmax_act17_par, minmax_act18_par, on='states', how='inner')
act1718_par_compare

states  act17_participation  act18_participation
0            Maine                 0.08                 0.07
1          Alabama                 1.00                 1.00
2             Utah                 1.00                 1.00
3        Tennessee                 1.00                 1.00
4   South Carolina                 1.00                 1.00
5         Oklahoma                 1.00                 1.00
6   North Carolina                 1.00                 1.00
7           Nevada                 1.00                 1.00
8          Montana                 1.00                 1.00
9         Missouri                 1.00                 1.00
10     Mississippi                 1.00                 1.00
11       Louisiana                 1.00                 1.00
12        Kentucky                 1.00                 1.00
13        Arkansas                 1.00                 1.00
14       Wisconsin                 1.00                 1.00
15         Wyoming                 1.00                 1.00

From the above, we observe the following:

- **2017 ACT**
    - **Lowest participation rates:** 1 state (Maine)
    - **Highest participation rates:** 17 states
- **2018 ACT**
    - **Lowest participation rates:** 1 state (Maine)
    - **Highest participation rates:** 17 states
- Maine consistently has the lowest ACT participation rates, and the rate drops from 8% (2017) to 7% (2018).
- 15 states consistenly have the highest SAT participation rates of 100%.

In [143]:
# Creating masks to obtain min & max values of 'sat17_total' column
min_sat17_total = satact_1718['sat17_total']==satact_1718['sat17_total'].min()
max_sat17_total = satact_1718['sat17_total']==satact_1718['sat17_total'].max()

# Applying the masks to obtain states with the lowest OR highest average total scores,
# and then sorting the rows by average total scores in ascending order.
minmax_sat17_total = satact_1718[['states', 'sat17_total']].loc[(min_sat17_total) | (max_sat17_total)].sort_values(by='sat17_total')
minmax_sat17_total

states  sat17_total
8   District of Columbia          950
23             Minnesota         1295

In [144]:
# Creating masks to obtain min & max values of 'sat18_total' column
min_sat18_total = satact_1718['sat18_total']==satact_1718['sat18_total'].min()
max_sat18_total = satact_1718['sat18_total']==satact_1718['sat18_total'].max()

# Applying the masks to obtain states with the lowest OR highest average total scores,
# and then sorting the rows by average total scores in ascending order.
minmax_sat18_total = satact_1718[['states', 'sat18_total']].loc[(min_sat18_total) | (max_sat18_total)].sort_values(by='sat18_total')
minmax_sat18_total

states  sat18_total
8   District of Columbia          977
23             Minnesota         1298

In [157]:
sat1718_total_compare = pd.merge(minmax_sat17_total, minmax_sat18_total, on='states', how='outer')
sat1718_total_compare

states  sat17_total  sat18_total
0  District of Columbia          950          977
1             Minnesota         1295         1298

From the above, we observe the following:

- **2017 SAT**
    - **Lowest average total score (950 / 1600) :** District of Columbia
    - **Highest average total score (1295 / 1600) :** Minnesota
- **2018 SAT**
    - **Lowest average total score (977 / 1600) :** District of Columbia
    - **Highest average total score (1298 / 1600) :** Minnesota
- District of Columbia consistently has the lowest average total score.
- Minnesota consistently has the highest average total score.

In [158]:
# Creating masks to obtain min & max values of 'act17_composite' column
min_act17_composite = satact_1718['act17_composite']==satact_1718['act17_composite'].min()
max_act17_composite = satact_1718['act17_composite']==satact_1718['act17_composite'].max()

# Applying the masks to obtain states with the lowest OR highest average total scores,
# and then sorting the rows by average total scores in ascending order.
minmax_act17_composite = satact_1718[['states', 'act17_composite']].loc[(min_act17_composite) | (max_act17_composite)].sort_values(by='act17_composite')
minmax_act17_composite

states  act17_composite
28         Nevada             17.8
29  New Hampshire             25.5

In [159]:
# Creating masks to obtain min & max values of 'act18_composite' column
min_act18_composite = satact_1718['act18_composite']==satact_1718['act18_composite'].min()
max_act18_composite = satact_1718['act18_composite']==satact_1718['act18_composite'].max()

# Applying the masks to obtain states with the lowest OR highest average total scores,
# and then sorting the rows by average total scores in ascending order.
minmax_act18_composite = satact_1718[['states', 'act18_composite']].loc[(min_act18_composite) | (max_act18_composite)].sort_values(by='act18_composite')
minmax_act18_composite

states  act18_composite
28       Nevada             17.7
6   Connecticut             25.6

In [160]:
act1718_comp_compare = pd.merge(minmax_act17_composite, minmax_act18_composite, on='states', how='outer')
act1718_comp_compare

states  act17_composite  act18_composite
0         Nevada             17.8             17.7
1  New Hampshire             25.5              NaN
2    Connecticut              NaN             25.6

From the above, we observe the following:

- **2017 ACT**
    - **Lowest average composite score (17.8 / 36.0) :** Nevada
    - **Highest average composite score (25.5 / 36.0) :** New Hampshire
- **2018 ACT**
    - **Lowest average composite score (17.7 / 36.0) :** Nevada
    - **Highest average composite score (25.6 / 36.0) :** Connecticut
- Nevada consistently has the lowest average composite score.

## Visualize the data

There's not a magic bullet recommendation for the right number of plots to understand a given dataset, but visualizing your data is *always* a good idea. Not only does it allow you to quickly convey your findings (even if you have a non-technical audience), it will often reveal trends in your data that escaped you when you were looking only at numbers.

Some recommendations on plotting:
- Plots have titles
- Plots have axis labels
- Plots have appropriate tick labels
- All text is legible in a plot
- Plots demonstrate meaningful and valid relationships
- Plots are interpreted to aid understanding

There is such a thing as too many plots, and there are a *lot* of bad plots. You might make some! (But hopefully not with the guided prompts below).

#### Use Seaborn's heatmap with pandas `.corr()` to visualize correlations between all numeric features

Heatmaps are generally not appropriate for presentations, and should often be excluded from reports as they can be visually overwhelming. **However**, they can be extremely useful in identify relationships of potential interest (as well as identifying potential collinearity before modeling).

*example*:
```python
sns.heatmap(df.corr())
```

Please take time to format your output, adding a title. Look through some of the additional arguments and options. (Axis labels aren't really necessary, as long as the title is informative).

#### Define a custom function to subplot histograms

We have data for two tests for two years. We only have composite (and not subtest scores) for the 2018 ACT. We should write a function that will take the names of 2+ columns and subplot histograms. While you can use pandas plotting or Seaborn here, matplotlib gives you greater control over all aspects of your plots.

[Helpful Link for Plotting Multiple Figures](https://matplotlib.org/users/pyplot_tutorial.html#working-with-multiple-figures-and-axes)

Here's some starter code:

In [ ]:
def subplot_histograms(dataframe, list_of_columns, list_of_titles, list_of_xlabels):
    nrows = int(np.ceil(len(list_of_columns)/2)) # Makes sure you have enough rows
    fig, ax = plt.subplots(nrows=nrows, ncols=2) # You'll want to specify your figsize
    ax = ax.ravel() # Ravel turns a matrix into a vector, which is easier to iterate
    for i, column in enumerate(list_of_columns): # Gives us an index value to get into all our lists
        ax[i].hist(dataframe[column]) # feel free to add more settings
        # Set titles, labels, etc here for each subplot

#### Plot and interpret histograms 
For each of the following:
- Participation rates for SAT & ACT
- Math scores for SAT & ACT
- Reading/verbal scores for SAT & ACT

In [ ]:
# Code

#### Plot and interpret scatter plots

For each of the following:
- SAT vs. ACT math scores for 2017
- SAT vs. ACT verbal/reading scores for 2017
- SAT vs. ACT total/composite scores for 2017
- Total scores for SAT 2017 vs. 2018
- Composite scores for ACT 2017 vs. 2018

Plot the two variables against each other using matplotlib or Seaborn

Your plots should show:
- Two clearly labeled axes
- A proper title
- Using colors and symbols that are clear and unmistakable

**Feel free to write a custom function, and subplot if you'd like.** Functions save both time and space.


In [ ]:
# Code

#### Plot and interpret boxplots

For each numeric variable in the dataframe create a boxplot using Seaborn. Boxplots demonstrate central tendency and spread in variables. In a certain sense, these are somewhat redundant with histograms, but you may be better able to identify clear outliers or differences in IQR, etc.

Multiple values can be plotted to a single boxplot as long as they are of the same relative scale (meaning they have similar min/max values).

Each boxplot should:
- Only include variables of a similar scale
- Have clear labels for each variable
- Have appropriate titles and labels

In [ ]:
# Code

#### Feel free to do additional plots below
*(do research and choose your own chart types & variables)*

Are there any additional trends or relationships you haven't explored? Was there something interesting you saw that you'd like to dive further into? It's likely that there are a few more plots you might want to generate to support your narrative and recommendations that you are building toward. **As always, make sure you're interpreting your plots as you go**.

#### (Optional): Using Tableau, create a choropleth map for each variable using a map of the US. 

Save this plot as an image file in an images directory, provide a relative path, and insert the image into notebook in markdown.

## Descriptive and Inferential Statistics

#### Summarizing Distributions

Above, we used pandas `describe` to provide quick summary statistics of our numeric columns. We also demonstrated many visual relationships.

As data scientists, having a complete understanding of data is imperative prior to modeling.

While we will continue to build our analytic tools, we know that measures of *central tendency*, *spread*, and *shape/skewness* provide a quick summary of distributions.

For each variable in your data, summarize the underlying distributions (in words & statistics)
 - Be thorough in your verbal description of these distributions.
 - Be sure to back up these summaries with statistics.

Answers:


#### Distributions in the data

In this dataset, each data represents a sample from a population.                        
For example, for ACT math test:
- Population: the test results of all the students who take this test, nation-wide.
- Population mean: is the national average of ACT math test (total scores/total no. of test takers) 
- Sample: the state means of ACT math test. We have 51 samples (51 states)

***According to CLT, we generally assuming that data we sample from a population will be normally distributed. Do we observe this trend?***

Answer:

Does This Assumption Hold for:
    - Math
    - Reading
    - Rates
Explain your answers for each distribution and how you think this will affect estimates made from these data.

Answer:

#### Estimate Limits of Data

Suppose we only seek to understand the relationship between SAT and ACT participation rates in 2017. 

##### Does it make sense to conduct statistical inference given these data specifically? 

Why or why not?

*(think about granularity, aggregation, the relationships between populations size & rates...consider the actually populations these data describe in answering this question)*

Answer:

##### Is it appropriate to compare *these* specific SAT and ACT math scores  - can we say students with higher SAT math score is better than those with lower ACT math score, or vice versa?

Why or why not?

Answer:

#### Statistical Evaluation of Distributions 

**If you feel it's appropriate**, using methods we discussed in class, run hypothesis tests to compare variables of interest in our dataset. 

In [ ]:
# Code:

## Outside Research

Based upon your observations, choose **three** states that demonstrate interesting trends in their SAT and/or ACT participation rates. Spend some time doing outside research on state policies that might influence these rates, and summarize your findings below. **Feel free to go back and create new plots that highlight these states of interest**. If you bring in any outside tables or charts, make sure you are explicit about having borrowed them. If you quote any text, make sure that it renders as being quoted. (Make sure that you cite your sources -- check with you local instructor for citation preferences).

## Conclusions and Recommendations

Based on your exploration of the data, what are you key takeaways and recommendations? Choose one state with a lower participation rate and provide a suggestion for how the College Board might increase participation amongst graduating seniors in this state. Are there additional data you desire that would better inform your investigations?